Enable use of psycopg2.execute_values()

ydinari opened the pull request:
<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->
Fixes: #4623
In SQLCompiler.visit_insert(), put logic to generate the VALUES clause into function (generate_values_placeholders_str) to enable overriding just that logic in derived classes.
In PGCompiler_psycopg2 override generate_values_placeholders_str() to enable use of psycopg2.execute_values() which is faster than execute_batch().
Extend flag use_batch_mode to indicate which psycopg2 method to use.
In PGDialect_psycopg2.do_executemany(), call the correct psycopg2 method based on the use_batch_mode flag and compiler variables.
Add tests for psycopg2.execute_values().

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

Commit:
Fixes: #4623
- In SQLCompiler.visit_insert(), put logic to generate the VALUES clause into function (generate_values_placeholders_str) to enable overriding just that logic in derived classes
- In PGCompiler_psycopg2 override generate_values_placeholders_str() to enable use of psycopg2.execute_values() which is faster than execute_batch()
- Extend flag use_batch_mode to indicate which psycopg2 method to use
- Add tests for psycopg2.execute_values()

Commit:
Change field name from multi_params to clearer multiple_rows
Fixes: #4623

zzzeek commented:
vonage, wow I am a customer in fact.

OK, this looks pretty good, it will need pep8 formatting fixes as there are tests that will fail right now without them.   Do you want me to show you how that can be mostly automated?  if you install [pre-commit](https://pre-commit.com/) it will run code formatting automatically, however there might be some long lines here you'd have to fix manually.

I'll run this through gerrit so you can see.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work to try to get revision b7f6fb5f522db7e03c9cce7bc21222aa0843d166 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change b7f6fb5f522db7e03c9cce7bc21222aa0843d166: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1355

Commit:
PEP8 changes created automatically by autopep8 tool
Fixes: #4623

Commit:
- Create new flag, execution_mode, to replace use_batch_mode, that more correctly reflects the controlled behavior
- When flag indicates using psycopg2.execute_values but query is not insert, use psycopg2.execute_batch (and not the slow psycopg2.executemany)
Fixes: #4623

Commit:
* Increase default page_size for calling psycopg2.execute_values() from 2,000 to 10,000, as experiments show better performance
* Fix documentation
Fixes: #4623

Commit:
Fix documentation

Commit:
* Fix usage of flag execution_mode
* Fix test to use new flag
Fixes: #4623

Commit:
Internally use an enum type for handling execution_mode option, and to enforce valid string value for the option
Fixes: #4623

Commit:
Remove debug printing

zzzeek commented:
I'm going to be out of town for a week so thanks very much for your patience.

Commit:
Take psycopg2.execute_values() page_size parameter value from a flag instead of having it hardcoded in PGCompiler_psycopg2
Fixes: #4623

Commit:
Add page_size flag test
Fixes: #4623

zzzeek commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -551,13 +590,55 @@ def _log_notices(self, cursor):
+    ):
+        self.dialect = dialect
+        self.multiple_rows = inline
+        self.execute_values_insert_template = None
```
this can be set as a class-level variable to start wtih, as None, that would allow you to not have to rewrite the whole constructor:

```
class PGCompiler_psycopg2(PGCompiler):
    execute_values_insert_template = None

```

ydinari replied:
OK. You mean I guess all those variables?

ydinari replied:
But what about the other two variables? I still need to override the constructor to get those

ydinari replied:
And also, is this class guaranteed to have only one instance? If not it's problematic to use a class variable to hold the template

zzzeek replied:
So there are the "other two variables", "dialect", and "multiple_rows", for which I don't see the purpose for these assignments.  The PGCompiler_psycopg2 object already has a .dialect instance variable you can access as "self.dialect".   The "inline" keyword argument is also available on the object as just that, "self.inline", there is no need to copy it to an additional instance variable.

For "execute_values_insert_template", it appears you only modify execute_values_insert_template in the generate_values_placeholders_str method.   You can safely default this to None at the class level.  When you later invoke "self.execute_values_insert_template = 'some value'", that sets an *instance level* variable which overrides the class-level default version.    SQLAlchemy uses this pattern all over the place:

```
Python 3.7.3 (default, May 11 2019, 00:38:04) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo:
...    some_arg = None
... 
>>> 
>>> f1 = Foo()
>>> f1.some_arg
>>> f2 = Foo()
>>> f2.some_arg = 'a'
>>> f3 = Foo()
>>> f3.some_arg = 'b'
>>> f1.some_arg
>>> f2.some_arg
'a'
>>> f3.some_arg
'b'

```


If I'm misreading or misinterpreting what I'm seeing here, that is very possible.    So let me know if I'm doing that here.

ydinari replied:
OK. I missed the dialect variable, and didn't know about this behavior of class variables.

ydinari replied:
As for multiple_rows, I wasn't sure that reading self.inline the correct way to get this info. But since you didn't comment about that, I guess it is :)

zzzeek commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -551,13 +590,55 @@ def _log_notices(self, cursor):
     pass
 
 
+class EnumPsycopg2ExecutionMode(Enum):
```
we support Python 2, can't use Enum.  Please use util.symbol:

```
from .. import util
SINGLE_STATEMENT = util.symbol("single_statement")
```

ydinari replied:
ok

zzzeek commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -610,6 +691,8 @@ def __init__(
         use_native_hstore=True,
         use_native_uuid=True,
+        execution_mode=None,
+        page_size=10000,
```
these names are both too generic if they are at the dialect level.    execution_mode needs to refer to "executemany" and "page_size" needs to refer to "execute_values".

ydinari replied:
OK.
page_size can be used for execute_batch as well. I didn't use it in this pull request to keep the notion that no change should be made when not using the new flags. However, it will improve the performance considerably. So, if we want to add it to the call to execute_batch(), maybe we should give the flag a more generic name (like executemany_page_size) ?

zzzeek replied:
just that it has "executemany" in the name should be enough.

ydinari replied:
ok

zzzeek commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -766,11 +861,21 @@ def on_connect(conn):
-            extras = self._psycopg2_extras()
-            extras.execute_batch(cursor, statement, parameters)
-        else:
+        if self.psycopg2_execution_mode == EnumPsycopg2ExecutionMode.SINGLE_STATEMENT:
```
use "is" for comparing symbols, not ==

ydinari replied:
ok

zzzeek commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -551,13 +590,55 @@ def _log_notices(self, cursor):
+    """
+    Enum type to describe psycopg2 execution mode of multiple parameters queries
+    """
+    SINGLE_STATEMENT = 'single_statement'
```
Can we use a more descriptive set of names here ?    I'm not sure what "single statement" describes without reading the source code.   In all three cases, we are invoking a "statement", and we are calling a method on psycopg2 only once, so from that perspective, they are all "single statement".  Within the internals of psycopg2 executemany(), the default implementation runs the given statement for each parameter set, so that's also not "single".

I think a set of parameters / names that are more descriptive might be:   "executemany_mode = (EXECUTEMANY_DEFAULT|EXECUTEMANY_BATCH|EXECUTEMANY_VALUES)".   that way they are all specific to executemany and they indicate which ones are special and also map directly to the names used by psycopg2; someone using this API will likely be referring to psycopg2s documentation.

ydinari replied:
ok

zzzeek reviewed (commented):
hi !   I see you've been working on this a lot, thanks!  I made a bunch of initial review comments.

Commit:
PR review fixes:
- Change flag and flag parameters names
- Use util.symbol instead of Enum for to support python2
Fixes: #4623

zzzeek commented:
send me a ping here when you want me to pull this through CI and review, thanks!

Commit:
- Change flag name execute_values_page_size to executemany_page_size
- Remove unnecessary constructor override in PGCompiler_psycopg2
Fixes: #4623

ydinari commented:
Hi, check out last commit! thanks

ydinari commented:
Hi,
Any progress with the review?
Thanks

zzzeek commented:
OK, i just looked and the changes look very nice, thanks for making them just like I said.   I still have to decide if I like those names, because they are a little cumbersome, but that's on me.   it may be that we remove the "psycopg2_" prefix perhaps.

we need to decide:

- it's easier to introduce this in 1.4 but that would mean it won't be available for a while.   This change is likely small enough to be in 1.3 but needs some verbiage around the switch from use_batch_mode to psycopg2_executemany_mode; this goes in the docs with a "..versionchanged" section and also I like to emit a deprecation warning for the old flag.   I have a loose policy that deprecations aren't introduced mid-minor release but as this is an "experimental" feature, a warning that the flag is "superseded" might not be so bad.

- if we do this in 1.3, in 1.4 we can perhaps default it to the highest performing option.    A new minor release is allowed to have bigger regressions when it first comes out (usually as betas, which nobody tests anyway) but that's a handy way for more dramatic behavioral changes to come out and get tested.

- I'm not sure if I saw test coverage for the ArgumentError part of this.  Also, I wonder if the pattern where they pass a string that gets converted to a symbol could somehow be a sqlalchemy.util function, and I feel like I want to look elsewhere in the codebase for this pattern too (oh, [it is](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/pool/base.py#L194) , maybe a separate commit can make that improvement separately)

I think if you can confirm you really were hoping to have this in 1.3 that would set me up for the next step.

Commit:
* Put back the test for old flag use_batch_mode
* Add tests of the arguments for executemany_mode and executemany_page_size flags
* Add check of executemany_page_size flag argument

ydinari commented:
Hi,
* Yes I really would like it to be in 1.3 so we can start using it soon.
* Let me know what you decide about using the flag by default or not so I'll know what to put in the versionchanged section, and also change the flag's default value if needed.
* I didn't understand what you meant about an sqlalchemy.util function to check for correct argument for the flag. The link shows code doing pretty much the same kind of checks I do, and there's no call to a util function.
* Arguments tests added

zzzeek commented:
I will try to get this is in 1.3.7.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work to try to get revision 715cfbea4d9f8f4e5ef239b0921a8dceb36ccbfb of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 715cfbea4d9f8f4e5ef239b0921a8dceb36ccbfb added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1355

Commit:
Shorten some lines to comply with PEP 8

zzzeek commented:
at this point it's faster if I take it into gerrit fully and add the remaining changelog notes and other doc tweaks etc.

zzzeek committed:
Rewrite pool reset_on_return parsing using a util function

Choosing a util.symbol() based on a user parameter is about to have
another use case added as part of #4623, so add a generalized solution
ahead of it.

Change-Id: I420631f81af2ffc655995b9cce9ff2ac618c16d7

zzzeek commented:
symbol parsing feature is in https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1405/

Commit:
* Use util.symbol.parse_user_argument() to parse executemany_mode option
* Correct test to check for error in the format util.symbol.parse_user_argument() issues
Fixes: #4623

ydinari commented:
Cherry picked the argument parsing commit and used it.
Are you doing the doc changes?
Thanks

zzzeek commented:
ill base this one in and take over, thanks

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work to try to get revision c3d3a36f7eb66c86d14ed9c1c31b4b48bd204855 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset c3d3a36f7eb66c86d14ed9c1c31b4b48bd204855 added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1355

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -52,11 +52,42 @@
-  It is currently experimental but
-  may well become True by default as it is critical for executemany
-  performance.
+.. _psycopg2_executemany_mode:
```
**mike bayer** (zzzeek) wrote:

I'm reworking the organization here since this is inside of a bulleted list stating all parameters,

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -551,13 +589,41 @@ def _log_notices(self, cursor):
+        # for receiving multiple return values from insert query
+        # Note: self.inline is true iff there are multiple parameters sets to
+        # the query
+        if self.inline and not returning_clause_exists and \
```
**mike bayer** (zzzeek) wrote:

is it guaranteed this will be an executemany()?

ydinari replied:
Do you mean when self.inline is true? This is what I deduced from the code, I hope that's right

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -551,13 +589,41 @@ def _log_notices(self, cursor):
+        # the query
+        if self.inline and not returning_clause_exists and \
+                self.dialect.psycopg2_executemany_mode is EXECUTEMANY_VALUES:
+            self.execute_values_insert_template = "(" + \
```
**mike bayer** (zzzeek) wrote:

this is the string we would generate anyway..right?   Why not have the base compiler just store this string in all cases if that's the case, then we don't need this method here.

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/psycopg2.py:
```diff
@@ -766,11 +867,22 @@ def on_connect(conn):
+                page_size=self.psycopg2_executemany_page_size)
+
+        else:  # EXECUTEMANY_VALUES of non-insert query, or EXECUTEMANY_BATCH
+            self._psycopg2_extras().execute_batch(cursor, statement,
```
**mike bayer** (zzzeek) wrote:

why aren't we using pagesize here?

ydinari replied:
We can, I didn't want to change the behavior for people not using the new flag.
If you want I'll add it

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

(5 comments)

Just some comments as I go through this.

* lib/sqlalchemy/dialects/postgresql/psycopg2.py (line 193): so the plan is that this gets deprecated since it's being superseded by a new system.

zzzeek commented:
hi -

to sum up changes I've made:

1. the biggest is that I think for now this is better as a short regular expression operation, which is done right in do_executemany(), rather than changing the base compiler class to farm out this hook.   It's a lot less code and doesn't add a funny looking concept to the base compiler that only matters for one specific DBAPI.  

2. I've changed the names of things, since we're in psycopg2 already we don't need "psycopg2_executemany_mode", it's just "executemany_mode".    The values then for this are None, "batch", and "values".

3. I've consolidated the unit tests into a single fixture, deprecated use_batch_mode

4. page_size applies to both execute_batch and execute_values

5. due to the above, I think we are safer deferring to psycopg2's defaults, even though 100 is quite low.  This is a power user feature so the docs clearly state that you should probably also set executemany_page_size too to a number like 10000 if you are using execute_values.

6. in an upcoming patch Im going to break page size into two variables for execute_batch and execute_values, as we have fallback logic here; I think 10000 is too high for execute_batch so these need to be separate

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

I think I'm happy with this now.  I'll try to merge tomorrow if I have time unless you have suggestions.

ydinari commented:
Hi, so where do I see the changes you made?
And do you need anything from me?

zzzeek commented:
> Hi, so where do I see the changes you made?
> And do you need anything from me?

oh, OK so this PR is mirrored in my gerrit server at https://gerrit.sqlalchemy.org/1413 and one of the main reasons I do that is so that I can also push to it, so it has moved past the PR now.   i just need you to flag anything that isn't going to work for you at this point.

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/1355 has been **merged**. Congratulations! :)

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/1413 has been **merged**. Congratulations! :)

ydinari commented:
Hi,
Great!
When will 1.3.7 be released?
Also, looking at the new implementation of do_executemany(), I see there is no check whether the query contains a RETURNING clause. Is that ok, because currently sqlalchemy doesn't generate insert statements with both multiple parameters and RETURNING?
BTW, what kind of account do you have with Vonage?
Thanks,
Yuval

zzzeek commented:
it's released.    Right now there aren't any codepaths that will automatically use a "returning" statement with executemany(), and the current behavior is that fetchone() doesn't work for these in any case since no DBAPI supports fetching rows after executemany.    I have a residential account with vonage.

ydinari commented:
Hi,
There is a problem with the regular expression used in do_executemany - it doesn't work if the query contains a newline character (which it does in our case).
Possible solutions is write it like this:
`_insert_values_match = re.compile(r".* VALUES (\(.+\))", re.DOTALL).match`
Or:
`_insert_values_search = re.compile(r" VALUES (\(.+\))").search`
Do you want me to open a new PR or will you do it?
Thanks

zzzeek commented:
pretty unfortunate timing!   as always, the test case is what's needed.

how are you creating the insert with the newline ?   it looks like if things like CTEs are there and such which could add newlines.

if you can send a PR that includes tests then I can get that in.

can we think of any scenario where the " VALUES ()" could appear twice in a single INSERT statement ?

zzzeek commented:
also can you confirm the current behavior is that it's degrading to execute_batch, right?  eg. it isn't failing, just isn't using the optimization.

zzzeek commented:
also, I will need you to fully sign off and test this *completely* in your environment before I can release again.    I don't usually put features this significant in a point release.  it is very risky.

zzzeek commented:
I've created new tests and revised the approach to be more like the approach you orgiinally had in https://gerrit.sqlalchemy.org/#/c/1419/ .   Pretty sure this one will work consistently.

ydinari commented:
Hi,
OK, I'll test it in our environment. On which branch should I be?
As for what you asked:
- In our every sql statement starts with some comment after which there's a newline. But in the general case I think sql statements may include newline anywhere.
- I don't know whether sql syntax allows multiple VALUES keyword in a single statement.
- Yes, the behavior is that execute_batch() is called (but then the executemany_values_page_size flag is ignored).

ydinari commented:
Hi,
It works in my environment.
There's an issue you decide if important or not: the debug log show the query as it is before the replacement of the VALUES clause, which can be confusing.
Thanks

zzzeek commented:
> Hi,
> It works in my environment.
> There's an issue you decide if important or not: the debug log show the query as it is before the replacement of the VALUES clause, which can be confusing.
> Thanks

you mean, it shows the regular INSERT query with a single VALUES clause, right ?  I guess earlier when we did this up at the compiler phase, it would show the modified query, but then it wouldn't show the template at all.  psycopg2 is not unique in this particular optimization, they just make it something the end-user has to do explicitly, if you are using for example mysqlclient, they are doing it [here](https://github.com/PyMySQL/mysqlclient-python/blob/master/MySQLdb/cursors.py#L21) and [here](https://github.com/PyMySQL/mysqlclient-python/blob/master/MySQLdb/cursors.py#L229), which IMHO is what psycopg2 should have done for us rather than burdening us with all this explicitness, but I can certainly relate to when a maintainer doesn't want to take on a complex new behavior.

anyway the point is that MySQL is doing this same thing and we can't see that either so we can likely get away with considering this to be a driver detail.

zzzeek commented:
as far as newlines, yes a SQL statement can have newlines, but the assumption I had was that these were INSERTs created by the compiler, which don't have newlines in them.  However, as my test in the patch demonstrates, adding a scalar SELECT as one of the values will put a newline right in the middle of the VALUES which makes the regexp approach not really work (even mysql's approach wouldn't work with that).

zzzeek commented:
and yes if you have an event handler adding comments, that too.   Is that how you are adding comments ?    The new patch includes a test for the case where someone's event handler has modified the query beyond where we know how to find the VALUES anymore.  this should not be the norm, but if an event handler really rewrote the whole query, it's going to degrade for now.

zzzeek commented:
OK I'm going to merge it..take two!

zzzeek commented:
after a few more tweaks, forgot to remove the regex..etc

ydinari commented:
In our code there's an `@event.listens_for(Engine, "before_cursor_execute", retval=True)` decorated method the prepend a comment with a newline to the existing statement.
So this will be release 1.3.8?

zzzeek commented:
yup 1.3.8 is next

ydinari commented:
Hi,
When will it be released?
Thanks

zzzeek commented:
was released a few days ago, all good?
